module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,640 @@
+	MaxCachedEntries uint64
+
+	// Max number of unsent notifications to allow before pausing new fetches. After exceeding this
+	// limit, the requester will stop processing new finalized block notifications. This prevents
+	// unbounded memory use by the requester if it gets stuck fetching a specific height.
+	MaxSearchAhead uint64
+
+	// The timeout for fetching ExecutionData from the db/network
+	FetchTimeout time.Duration
+
+	// Exponential backoff settings for download retries
+	RetryDelay    time.Duration
+	MaxRetryDelay time.Duration
+
+	// Whether or not to run datastore check on startup
+	CheckEnabled bool
+}
+
+type executionDataRequester struct {
+	component.Component
+	cm      *component.ComponentManager
+	ds      datastore.Batching
+	bs      network.BlobService
+	eds     state_synchronization.ExecutionDataService
+	metrics module.ExecutionDataRequesterMetrics
+	log     zerolog.Logger
+
+	// Local db objects
+	blocks  storage.Blocks
+	results storage.ExecutionResults
+
+	status *status.Status
+	config ExecutionDataConfig
+
+	// Notifiers for queue consumers
+	blockNotifier engine.Notifier
+
+	// Job queues
+	blockConsumer        *jobqueue.WrappedConsumer
+	notificationConsumer *jobqueue.WrappedConsumer
+
+	// List of callbacks to call when ExecutionData is successfully fetched for a block
+	consumers []state_synchronization.ExecutionDataReceivedCallback
+
+	consumerMu sync.RWMutex
+}
+
+var _ state_synchronization.ExecutionDataRequester = (*executionDataRequester)(nil)
+
+// New creates a new execution data requester component
+func New(
+	log zerolog.Logger,
+	edrMetrics module.ExecutionDataRequesterMetrics,
+	datastore datastore.Batching,
+	blobservice network.BlobService,
+	eds state_synchronization.ExecutionDataService,
+	processedHeight storage.ConsumerProgress,
+	processedNotifications storage.ConsumerProgress,
+	state protocol.State,
+	blocks storage.Blocks,
+	results storage.ExecutionResults,
+	cfg ExecutionDataConfig,
+) (state_synchronization.ExecutionDataRequester, error) {
+	e := &executionDataRequester{
+		log:           log.With().Str("component", "execution_data_requester").Logger(),
+		ds:            datastore,
+		bs:            blobservice,
+		eds:           eds,
+		metrics:       edrMetrics,
+		blocks:        blocks,
+		results:       results,
+		config:        cfg,
+		blockNotifier: engine.NewNotifier(),
+	}
+
+	executionDataNotifier := engine.NewNotifier()
+	rootHeight := e.config.StartBlockHeight - 1
+
+	var err error
+	e.blockConsumer, err = jobqueue.NewWrappedConsumer(
+		log.With().Str("module", "block_consumer").Logger(),
+		processedHeight,
+		status.NewSealedBlockReader(state, blocks),
+		e.processBlockJob,
+		e.blockNotifier.Channel(),
+		rootHeight,
+		jobqueue.WithMaxProcessing(fetchWorkers),
+		// notifies notificationConsumer when new ExecutionData blobs are available
+		jobqueue.WitNotifier(func(module.JobID) { executionDataNotifier.Notify() }),
+	)
+	if err != nil {
```

Comment:
```suggestion
	// create a worker that process a single job of downloading execution data for a sealed block by height
	// and report its completion 
	sealedBlockWorker := NewReadyDoneAwareWorker(
	  e.processBlockJob, // process the job of a sealed block to download its execution data
	  func(blockHeight) { finalizationConsumer.NotifyJobIsDone(blockHeight) }, // report job completion
	)
	
	// create a consumer that listens to the finalization events, and check by there is newly sealed blocks.
	// If there is, then create jobs and notify at most `fetchWorkers` number of workers to process them.
	// when a sealed block's execution data has been downloaded, it persists the height as last downloaded
	// if and only if its previous height has been downloaded and persisted.
	finalizationConsumer := NewReadyDoneAwareConsumer(
	  log, 
	  NewSealedBlockReader(state, blocks), // read sealed block by height
	  e.progress, // read and persist the downloaded height
	  e.finalizationNotifier.Channel(), // to listen to finalization events in order to find newly sealed blocks
	  sealedBlockWorker, // the worker that downloads for a specific sealed block and notify job completion.
	  fetchWorkers, // the number of concurrent workers
	)
	
	...
	func (e) OnBlockFinalized(*model.Block) {
		e.finalizationNotifier.Notify()
	}
```


I couple of suggestions here:
1. I think we only need one consumer and one worker to download the execution data in an async manner, as explained in the code and the comments.
2. If we want to notify when a block's execution data has been downloaded, we can let the sealedBlockWorker call some function when job is complete.
3. I don't think we need the `status` module. To me, it's a duplicated functionality of the job queue/consumer itself. The  `module/jobqueue/consumer` has taken care of all the stateful part of work. For instance:
a) the job queue [keeps track of the processing jobs](https://github.com/onflow/flow-go/blob/petera/execution-data-requester/module/jobqueue/consumer.go#L42)
b) the worker doesn't need to [spin off new workers,](https://github.com/onflow/flow-go/blob/petera/execution-data-requester/module/jobqueue/wrappedworker.go#L54-L59) because the [job consumer has done that already](https://github.com/onflow/flow-go/blob/petera/execution-data-requester/module/jobqueue/consumer.go#L239-L245).

Reply:
one of the constraints is that the requester should notify its event consumers about new execution data downloads in block order. So if there's more than one fetch worker, there needs to be some additional synchronization that tracks downloads and sends notifications as they are ready.

I think you're right that the status module could be removed or refactored to a much smaller implementation. Specifically, the heap isn't really necessary anymore. I left it around to simplify the caching, but that could be keyed off the `processedIndex` instead.

I think the worker does need to spin off the workers to support the ReadyDoneAware interface, at least until `jobqueue.Consumer` can be updated to pass a context to `Run`. We need some way to pass the irrecoverable context into the worker's run method without passing it in via `Run`. This PR implements it similarly to how it's done for [network.Register](https://github.com/onflow/flow-go/blob/master/network/p2p/network.go#L257-L257). I'm definitely open to ideas for how to do this in a simpler way.

Reply:
>one of the constraints is that the requester should notify its event consumers about new execution data downloads in block order.

How is this being done right now?

Reply:
the notification consumer uses `NextNotification` to get the notification to send. That method only returns a value if the next height is available, otherwise it waits.

I'm refactoring that bit to remove the heap now that the jobqueue handles the state management. that should be in soon

Reply:
As of faaa0ff, the `status.Status` struct is the reader for notifications. Its `AtIndex` either returns the next block height to notify for if it's available, or `storage.NotFound`. It maintains a map of available blocks, which is updated as new blocks are downloaded and notifications are sent